nomad/keyring_endpoint.go

```diff
@@ -360,3 +360,65 @@ func (k *Keyring) Delete(args *structs.KeyringDeleteRootKeyRequest, reply *struc
 	reply.Index = index
 	return nil
 }
+
+// ListPublic signing keys used for workload identities. This RPC is used to
+// back a JWKS endpoint.
+//
+// Unauthenticated.
+func (k *Keyring) ListPublic(args *structs.GenericRequest, reply *structs.KeyringListPublicResponse) error {
+
+	// JWKS is a public endpoint: intentionally ignore auth errors and only
+	// authenticate to measure rate metrics.
+	k.srv.Authenticate(k.ctx, args)
+	if done, err := k.srv.forward("Keyring.ListPublic", args, args, reply); done {
+		return err
+	}
+
+	defer metrics.MeasureSince([]string{"nomad", "keyring", "list_public"}, time.Now())
+
+	// Expose root_key_rotation_threshold so consumers can determine reasonable
+	// cache settings.
+	reply.RotationThreshold = k.srv.config.RootKeyRotationThreshold
+
+	// Setup the blocking query
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, s *state.StateStore) error {
+
+			// retrieve all the key metadata
+			snap, err := k.srv.fsm.State().Snapshot()
+			if err != nil {
+				return err
+			}
+			iter, err := snap.RootKeyMetas(ws)
+			if err != nil {
+				return err
+			}
+
+			pubKeys := []*structs.KeyringPublicKey{}
+			for {
+				raw := iter.Next()
+				if raw == nil {
+					break
+				}
+
+				keyMeta := raw.(*structs.RootKeyMeta)
+				if keyMeta.Inactive() {
+					// Skip inactive keys
+					continue
+				}
```

Comment:
I went back to the RFC on this and it's not totally clear to me how this will work with root key rotation and identities being pulled from the client asynchronously. Suppose the following scenario:

* Allocation is placed with identity signed with root key A
* Root key is rotated to key B
* Allocation makes a request to third party with identity signed by A
* Third-party hasn't seen this key before, and makes the JKWS request but the public key A is missing

I think we still need to show the public key for any key that can possibly have an unexpired JWT? If we cap the maximum JWT expiry to the root key rotation window (30d), that would suggest we need to display the active key + at least one previous key.

(I might suggest there's no harm in showing _all_ the inactive keys but because the root key is used to encrypt Variables too and we can't GC them without a full rekey, his could be a fairly large set over time.)

Reply:
> I went back to the RFC on this and it's not totally clear to me how this will work

Yeah when writing the RFC I assumed `inactive` meant *invalid* and that is not the case: `inactive` keys are valid.

When fixing this up I was going to switch this logic to only skip `deprecated` keys, but it appears that state is deprecated: https://github.com/hashicorp/nomad/blob/v1.6.1/nomad/structs/keyring.go#L61-L64

JWKS must include the key for any credential that should be considered valid.

JWKS should not include keys for any credentials that should already be expired.

I can think of 2 paths forward, but both will require a new `pending` key status:

#### `pending` keys

Where "creation" is cluster creation (or at least when they first upgrade).

1. At `creation + rotation_threshold / 2` the leader should create a new key with the `pending` status. Pending keys must never be used for signing!
2. At `creation + rotation_threshold` the current `active` key becomes `inactive`, and the current `pending` key becomes `active`.
3. For `rotation_threshold / 2` amount of time there is no `pending` key (we want to limit how early we create it a bit because the longer a root key exists, the more likely it is to get leaked).
4. Goto 1

This will be our pre-publishing mechanism like Vault implemented in https://github.com/hashicorp/vault/pull/12414 (although the description makes it sound caching related, the code is pre-creating the next key).

#### A. Redefine Key GC

This approach adjusts our existing `root_key_gc_threshold` to account for JWT's new expiration behavior:

1. Maintain the existing [`root_key_rotation_threshold = "30 days"`](https://developer.hashicorp.com/nomad/docs/configuration/server#root_key_rotation_threshold)
6. Change [`root_key_gc_threshold`](https://developer.hashicorp.com/nomad/docs/configuration/server#root_key_gc_threshold) to `30 days` from `1 hour`.
7. If `gc_threshold < rotation_threshold`, then force `gc_threshold = rotation_threshold` and spew a big warning in server logs.
8. Validate that `identity.ttl <= gc_threshold`.

This means the JWKS endpoint can expose all non-deprecated keys (and if someday deprecated keys go away, so can that `if` statement here).

Users who manually configured their `rotation_threshold > root_key_gc_threshold` will get the `gc_threshold` forcibly bumped on upgrade and a warning in their logs. This sucks as it's changing behavior that a user specifically selected, but I think it's our best choice because:

1. JWT expiration fundamentally impacts how the keyring is used, so I think it's appropriate to adapt the keyring's operation to properly handle that.
2. _Functionally_ there should be no user visible changes: variables will still work as before. JWTs will work as before. `nomad operator root keyring rotate` will still work. `inactive` keys will just exist for much longer.

#### B. New Key Status

Instead of changing the GC threshold we could add an `invalid` or `expired` status for keys in addition to the new `pending` status both approaches require. (Simply un-deprecating `deprecated` might also be sufficient. I'm unclear on the plan/story there.)

1. When a key is rotated its status still becomes `inactive`. These keys are still *valid* and therefore included in this endpoint.
2. After `rotation_threshold` passes the key becomes `invalid` (or whatever name we choose).  These keys are *invalid* and therefore excluded from this endpoint.
3. After `gc_threshold` elapses we GC *invalid* keys, not *inactive* keys.

The pro here is that we don't change anyone's settings out from underneath them...

...the con is that we do that by just injecting a completely new intermediate state they have no control over. :sweat_smile: 

Functionally the approaches seem to the same to me. I think the new status (B) might be more user friendly as it makes invalid keys *observable* instead of gc'ing them away for good. However, new states offer opportunities for new bugs! If we forget that we now store *untrusted* invalid keys and accidently use one somewhere, that's a CVE. :cry:

Reply:
Some additional background to consider when chosing between these two options:
* Because it's such an expensive operation we don't currently, by default, ever re-key Variables. The periodic rotation just rotates the key for new Variables we encrypt, and even `nomad operator root keyring rotate` requires a `-full` flag to kick off a re-key. So keys aren't going to get GC'd unless an org's Variables are ephemeral or get frequently updated (the updates will get encrypted with the active key).
* The original intent of `deprecated` was that after all Variables for a key were re-keyed and Workload Identities expired, the key could be safely deleted manually. I can't find in the RFC or the original PRs why we didn't simply delete the key at that point but it might have been paranoia/debuggability.
* We intentionally punted on trying to figure out a better solution to the full re-key because as you noted at the time a 30d rotation means we end up with 120 keys after 10 _years_, and that's just not enough to care about.

The `pending` key state sounds great. The `rotation_threshold / 2` does mean we're publishing 15 days in advance by default, which seems like a lot, but that's a nitpick and I'm fine with it if Vault was.

Option **A. Redefine Key GC** changes the _threshold_ but otherwise leaves the existing GC behavior intact. That includes not re-keying by default and leaving the keys around for potentially quite a long while.

Option **B. New Key Status** is arguably rolling back to what we were originally trying to do with the `deprecated` state, except with a working and reasonable expiration policy on Allocations because the JWTs expire now instead of living the lifetime of the Allocation. But the key can't really be considered `deprecated` until we've done a re-key, and that makes its value more dubious.

A third option would be to have an `expired` state (same as Option B) and when the GC finds keys in the expired state it kicks off a very slow re-key of all the Variables using that key, and then the `expired` key can be deleted. But we could also do the exact same thing with the `inactive` state by checking the rotation threshold first to ensure no JWTs are using the key.

So all that being said and one option not being "obviously" better than another, I'd advocate for option A with a follow-up down the road to do a slow re-key of Variables using `inactive` keys that are older than the GC threshold.

Reply:
> The pending key state sounds great. The `rotation_threshold / 2` does mean we're publishing 15 days in advance by default, which seems like a lot, but that's a nitpick based only on vibes and I'm fine with it if Vault was.

Good point. I think we'll probably want probably want to see if we can keep rotations on the order of single-digit-days.

To be clear the `/ 2` was just a nice round number to go for. My goal was to keep it as simple, proportional, and deterministic as possible without just forcing every little decision on users via config parameters. We could really easily put users in the position of:

> Ok so this alloc's client disconnected 3 days ago, has a max disconnect of 6 days, root keys are rotated every 14 days, and my identities have ttl's of 4 days so.........

That scenario is like 4 time-based config parameters already and doesn't cover JWKS or the behavior of 3rd party systems... so I kind of hope to just keep renewal logic proportional.

All that to say I'll make a note to revisit this when the server rotation PR goes up.

> I'd advocate for option A with a follow-up down the road to do a slow re-key of Variables using inactive keys that are older than the GC threshold.

Thanks for all of the context. It is interesting how distinct the root Variable symmetric key is from the root Workload Identity asymmetric key... we could also "just" decouple those since Variables are stateful and Workload Identities are not (soon). As usual "just" is papering over the huge amount of work and complexity that would entail, so I doubt that's the right option.

That being said leaving keys-used-for-Variables-but-not-for-valid-WIs around indefinitely seems fine as long as we prune them from JWKS reasonably.

So yeah, I think we have some wiggle room here both in initial (upcoming!) implementation as well as how to evolve it over time.